embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/telemetry/EmbraceTelemetryService.kt

```diff
@@ -0,0 +1,77 @@
+package io.embrace.android.embracesdk.telemetry
+
+import io.embrace.android.embracesdk.capture.metadata.EmbraceMetadataService
+import io.embrace.android.embracesdk.internal.spans.toEmbraceAttributeName
+import java.util.Locale
+
+/*
+    Service for tracking usage of public APIs, and different internal metrics about the app.
+ */
+internal class EmbraceTelemetryService(
+    private val okHttpReflectionFacade: OkHttpReflectionFacade = OkHttpReflectionFacade()
+) {
+
+    private val usageCountMap = mutableMapOf<String, Int>()
+    private val appAttributesMap = mutableMapOf<String, String>()
+
+    /*
+        Tracks the usage of a public API by name. Adds a suffix for easier identification. Replaces whitespaces with underscores.
+    */
+    fun onPublicApiCalled(name: String) {
+        val suffixedName = "emb.usage.${
+        name.toLowerCase(Locale.ENGLISH)
+            .trim()
+            .replace(" ", "_")
+        }"
+        usageCountMap[suffixedName] = (usageCountMap[suffixedName] ?: 0) + 1
+    }
+
+    /*
+        Returns a map with every telemetry value. For now, it's just usage counts of public APIs.
+     */
+    fun getTelemetryAttributes(): Map<String, String> {
+        val telemetryMap = mutableMapOf<String, String>()
+
+        telemetryMap.putAll(getUsageCountTelemetry())
+
+        telemetryMap.putAll(getAppAttributes())
+
+        return telemetryMap
+    }
+
+    private fun getUsageCountTelemetry() = usageCountMap.mapValues {
+        it.value.toString()
+    }.also {
+        usageCountMap.clear()
+    }
+
+    /**
+     * Interesting attributes about the running app environment. These should be the same for every session, so we only compute them once.
```

Comment:
IMO it'd be better to compute the OkHttp/app attributes as a `lazy` property. That way the functions won't mutate any state so will be easier to test & we can be sure it will behave ok if this is called concurrently.

Reply:
Yeah this whole thing can be a lazy/singleton as it is immutable. Now that are are calling this multiple times for an app launch, we should optimize this.

Reply:
In this case, lazy would work exactly like checking if it's initialized before computing, but it gives us concurrency safety, right?

Reply:
Yep